libopenage/curve/array.h

```diff
@@ -0,0 +1,244 @@
+
+
+	/**
+	 * Copy keyframes from another container to this container.
+	 *
+	 * Replaces all keyframes beginning at t >= start with keyframes from \p other.
+	 *
+	 * @param other Curve that keyframes are copied from.
+	 * @param start Start time at which keyframes are replaced (default = -INF).
+	 *              Using the default value replaces ALL keyframes of \p this with
+	 *              the keyframes of \p other.
+	 */
+	void sync(const Array<T, Size> &other, const time::time_t &start);
+
+
+	/**
+	 * Get the identifier of this curve.
+	 *
+	 * @return Identifier.
+	 */
+	size_t id() const override {
+		return this->_id;
+	}
+
+	/**
+	 * Get the human-readable identifier of this curve.
+	 *
+	 * @return Human-readable identifier.
+	 */
+	std::string idstr() const override {
+		if (this->_idstr.size() == 0) {
+			return std::to_string(this->id());
+		}
+		return this->_idstr;
+	}
+
+
+	KeyframeContainer<T>& operator[] (size_t index)
+	{
+		return this->container[index];
+	}
+
+	KeyframeContainer<T> operator[] (size_t index) const
+	{
+		return this->container[index];
+	}
+
+	class Iterator {
+	public:
+		Iterator(Array<T, Size> *curve, const time::time_t &time = time::TIME_MAX, size_t offset = 0) :
+			curve(curve), time(time), offset(offset) {};
+
+		const T operator*() {
+			return curve->frame(this->time, this->offset).second;
+		}
+
+		void operator++() {
+			this->offset++;
+		}
+
+		bool operator!=(const Array<T, Size>::Iterator &rhs) const {
+			return this->offset != rhs.offset;
+		}
+
+
+	private:
+		size_t offset;
+		Array<T, Size> *curve;
+		time::time_t time;
+	};
+
+
+	Iterator begin(const time::time_t &time = time::TIME_MAX);
+
+	Iterator end(const time::time_t &time = time::TIME_MAX);
+
+
+private:
+	std::array<KeyframeContainer<T>, Size> container;
+
+	//hint for KeyframeContainer operations
+	mutable std::array<size_t, Size> last_element = {};
+
+	/**
+	* Identifier for the container
+	*/
+	const size_t _id;
+
+	/**
+	* Human-readable identifier for the container
+	*/
+	const std::string _idstr;
+
+	/**
+	* The eventloop this curve was registered to
+	*/
+	const std::shared_ptr<event::EventLoop> loop;
+};
+
+
+template <typename T, size_t Size>
+std::pair<time::time_t, T> Array<T, Size>::frame(const time::time_t &t, const size_t index) const {
+	size_t frmae_index = container[index].last(t, this->last_element[index]);
+	this->last_element[index] = frmae_index;
+	return container[index].get(frmae_index).make_pair();
+}
+
+template <typename T, size_t Size>
+std::pair<time::time_t, T> Array<T, Size>::next_frame(const time::time_t &t, const size_t index) const {
+	size_t frmae_index = container[index].last(t, this->last_element[index]);
+	this->last_element[index] = frmae_index;
+	return container[index].get(frmae_index + 1).make_pair();
+}
+
+template <typename T, size_t Size>
+T Array<T, Size>::get(const time::time_t &t, const size_t index) const {
+	return this->frame(t, index).second;
```

Comment:
You should not call `frame` her because creating the `std::pair` is expensive and might prevent return value optimization. It's better to extract the value directly, evenm if that results in more lines of code.